adapter/src/main/kotlin/org/javacs/ktda/jdi/launch/KDACommandLineLauncher.kt

```diff
@@ -0,0 +1,208 @@
+    }
+
+    override fun defaultArguments(): Map<String, Connector.Argument> {
+        return this.defaultArguments
+    }
+
+    override fun toString(): String {
+        return name()
+    }
+
+    override fun transport(): Transport {
+        return transport
+    }
+
+    protected fun getOrDefault(arguments: Map<String, Connector.Argument>, argName: String): String {
+        return arguments[argName]?.value() ?: defaultArguments[argName]?.value() ?: ""
+    }
+
+    /**
+     * A customized method to launch the vm and connect to it, supporting cwd and env variables
+     */
+    @Throws(IOException::class, IllegalConnectorArgumentsException::class, VMStartException::class)
+    override fun launch(arguments: Map<String, Connector.Argument>): VirtualMachine {
+        val vm: VirtualMachine
+
+        val home = getOrDefault(arguments, ARG_HOME)
+        val options = getOrDefault(arguments, ARG_OPTIONS)
+        val main = getOrDefault(arguments, ARG_MAIN)
+        val suspend = getOrDefault(arguments, ARG_SUSPEND).toBoolean()
+        val quote = getOrDefault(arguments, ARG_QUOTE)
+        var exe = getOrDefault(arguments, ARG_VM_EXEC)
+        val cwd = getOrDefault(arguments, ARG_CWD)
+        val envs = urlDecode(getOrDefault(arguments, ARG_ENVS))?.toTypedArray()
+
+        check(quote.length == 1) {"Invalid length for $ARG_QUOTE: $quote"}
+        check(!options.contains("-Djava.compiler=") ||
+                options.toLowerCase().contains("-djava.compiler=none")) { "Cannot debug with a JIT compiler. $ARG_OPTIONS: $options"}
+
+        val listenKey = transportService?.startListening() ?: throw IllegalStateException("Failed to do transportService.startListening()")
+        val address = listenKey.address()
+
+        try {
+            val command = StringBuilder()
+
+            exe = if (home.isNotEmpty()) Paths.get(home, "bin", exe).toString() else exe
+            command.append(wrapWhitespace(exe))
+
+            command.append(" $options")
+
+            //debug options
+            command.append(" -agentlib:jdwp=transport=${transport.name()},address=$address,server=n,suspend=${if (suspend) 'y' else 'n'}")
+
+            command.append(" $main")
+
+            LOG.debug("command before tokenize: $command")
+
+            vm = launch(commandArray = CommandLineUtils.translateCommandline(command.toString()), listenKey = listenKey,
+                    ts = transportService!!, cwd = cwd, envs = envs
+            )
+
+        } finally {
+            transportService?.stopListening(listenKey)
+        }
+        return vm
+    }
+
+    internal fun wrapWhitespace(str: String): String {
+       return if(str.contains(' ')) "\"$str\" " else str
+    }
+
+    @Throws(IOException::class, VMStartException::class)
+    fun launch(commandArray: Array<String>,
+                        listenKey: TransportService.ListenKey,
+                        ts: TransportService, cwd: String, envs: Array<String>? = null): VirtualMachine {
```

Comment:
I think this would read better if we moved the parameters to individual lines:

```suggestion
    fun launch(
        commandArray: Array<String>,
        listenKey: TransportService.ListenKey,
        ts: TransportService,
        cwd: String,
        envs: Array<String>? = null
    ): VirtualMachine {
```